models/classes/accessControl/ActionAccessControl.php

```diff
@@ -0,0 +1,176 @@
+use oat\oatbox\service\ConfigurableService;
+use common_session_SessionManager as SessionManager;
+
+class ActionAccessControl extends ConfigurableService
+{
+    public const SERVICE_ID = 'tao/ActionAccessControl';
+
+    /**
+     * @example [
+     *     'controller' => [
+     *         'action1' => [
+     *             'role1' => 'READ',
+     *             'role2' => 'WRITE,
+     *         ],
+     *     ],
+     * ]
+     */
+    public const OPTION_PERMISSIONS = 'permissions';
+
+    public const READ = 'READ';
+    public const WRITE = 'WRITE';
+    public const GRANT = 'GRANT';
+
+    /**
+     * @example $permissionsToAdd = [
+     *     'controller' => [
+     *         'action1' => [
+     *             'role1' => 'READ',
+     *             'role2' => 'WRITE,
+     *         ],
+     *     ],
+     * ]
+     */
+    public function addPermissions(array $permissionsToAdd = []): void
+    {
+        $permissions = $this->getOption(self::OPTION_PERMISSIONS, []);
+
+        foreach ($permissionsToAdd as $controller => $actions) {
+            if (empty($permissions[$controller])) {
+                $permissions[$controller] = $actions;
+
+                continue;
+            }
+
+            foreach ($actions as $action => $rolePermissions) {
+                if (empty($permissions[$controller][$action])) {
+                    $permissions[$controller][$action] = $rolePermissions;
+
+                    continue;
+                }
+
+                $actionPermissions = $permissions[$controller][$action] ?? [];
+                $permissions[$controller][$action] = array_merge($actionPermissions, $rolePermissions);
+            }
+        }
+
+        $this->setOption(self::OPTION_PERMISSIONS, $permissions);
+    }
+
+    /**
+     * @example $permissionsToRemove = [
+     *     'controller' => [
+     *         'action1' => [
+     *             'role1',
+     *             'role2',
+     *         ],
+     *     ],
+     * ]
+     */
+    public function removePermissions(array $permissionsToRemove = []): void
+    {
+        $permissions = $this->getOption(self::OPTION_PERMISSIONS, []);
+
+        foreach ($permissionsToRemove as $controller => $actions) {
+            foreach ($actions as $action => $roles) {
+                foreach ($roles as $role) {
+                    unset($permissions[$controller][$action][$role]);
+                }
+
+                if (empty($permissions[$controller][$action])) {
+                    unset($permissions[$controller][$action]);
+                }
+            }
+
+            if (empty($permissions[$controller])) {
+                unset($permissions[$controller]);
+            }
+        }
+
+        $this->setOption(self::OPTION_PERMISSIONS, $permissions);
+    }
+
+    public function hasReadAccess(string $controller, string $action, ?User $user = null): bool
+    {
+        return $this->hasAccess([self::READ, self::WRITE, self::GRANT], $controller, $action, $user);
```

Comment:
Should not be like this?

```suggestion
        return $this->hasAccess([self::READ], $controller, $action, $user);
```

Reply:
Nope.
A user with WRITE access should be able to READ - he needs to view data which he wants to change.
Same with GRANT.

Reply:
But the method is `hasReadAccess`

Reply:
It's correct. If the user has WRITE access - he is also able to READ, so he has READ access

Reply:
So you are saying that if I want to check is I have READ access, I should have `GRANT`, 'READ' or `WRITE`?

I think I got it now :), For the scope we have now it is OK, but we will have different permission for NMC like `UPDATE`, 'CREATE' and 'DELETE' so we will have to change this logic here. Cause the permission will need to be exact the same.

Reply:
If it will be required - it can easily fixed 🙂 
Just change a bit of method declaration (first parameter `array` -> `string`)
and change comparison: `in_array($permission, $allowedPermissions, true)` -> `$permissions === allowedPermission`